polonius-engine/src/output/datafrog_opt.rs

```diff
@@ -299,11 +280,14 @@ pub(super) fn compute<T: FactTypes>(
 
             // subset(origin1, origin3, point2) :-
             //   live_to_dying_regions(origin1, origin2, point1, point2),
-            //   dying_can_reach_live(origin2, origin3, point1, point2).
-            subset_o1p.from_join(
+            //   dying_can_reach_live(origin2, origin3, point1, point2),
+            //   origin1 != origin3.
+            subset_o1p.from_join_filtered(
                 &live_to_dying_regions_o2pq,
                 &dying_can_reach_live,
-                |&(_origin2, _point1, point2), &origin1, &origin3| ((origin1, point2), origin3),
+                |&(_origin2, _point1, point2), &origin1, &origin3| {
+                    (origin1 != origin3).then(|| ((origin1, point2), origin3))
```

Comment:
Personally not a big fan of `bool::then`, but I don't mind if you and the others like it.

Reply:
I quite like it. If you and/or Niko find it too obtuse, we can change it of course.